pkg/distributor/distributor.go

```diff
@@ -2789,6 +2791,41 @@ respsLoop:
 	return result, nil
 }
 
+// adjustRequestLimitForReplicationSets recalculated the limit with respect to replication sets.
+// The returned value is the approximation, a query to an individual instance in the replication sets needs to be limited with.
+func adjustRequestLimitForReplicationSets(ctx context.Context, d *Distributor, replicationSets []ring.ReplicationSet, limit int) int {
+	if limit == 0 {
+		return limit
+	}
+
+	var shardSize int
+	if d.cfg.IngestStorageConfig.Enabled {
+		// When ingest storage is enabled, each partition, represented by one replication set is owned by only one ingester.
+		// So we use the number of replication sets to count the number of shards.
+		shardSize = len(replicationSets)
```

Comment:
`len(replicationSets)` includes both ACTIVE and INACTIVE partitions. To be more accurate, I think we should only consider ACTIVE partitions (the shard size on the write path, where series are written, is made only of ACTIVE partitions). It may not be a big deal: if we decide to not address it, then I would at least leave a comment here about it.